cmd/algorand-indexer/daemon.go

```diff
@@ -60,265 +64,336 @@ var (
+		return err
+	}
+	autoloadParamConfigPath := filepath.Join(cfg.indexerDataDir, autoLoadParameterConfigName)
+	paramConfigFound := util.FileExists(autoloadParamConfigPath)
+	// If we auto-loaded configs but a user supplied them as well, we have an error
+	if paramConfigFound {
+		if cfg.suppliedAPIConfigFile != "" {
+			err = fmt.Errorf("api parameter configuration was found in data directory (%s) as well as supplied via command line.  Only provide one",
+				filepath.Join(cfg.indexerDataDir, autoLoadParameterConfigName))
+			logger.WithError(err).Errorf("indexer parameter config error: %v", err)
+			return err
 		}
+		cfg.suppliedAPIConfigFile = autoloadParamConfigPath
+		logger.Infof("Auto-loading parameter configuration file: %s", suppliedAPIConfigFile)
+	}
+	return err
+}
 
-		// Detect the various auto-loading configs from data directory
-		indexerConfigFound := util.FileExists(filepath.Join(indexerDataDir, autoLoadIndexerConfigName))
-		paramConfigFound := util.FileExists(filepath.Join(indexerDataDir, autoLoadParameterConfigName))
+func createIndexerPidFile(pidFilePath string) error {
+	var err error
+	logger.Infof("Creating PID file at: %s\n", pidFilePath)
+	fout, err := os.Create(pidFilePath)
+	if err != nil {
+		err = fmt.Errorf("%s: could not create pid file, %v", pidFilePath, err)
+		logger.Error(err)
+		return err
+	}
+	_, err = fmt.Fprintf(fout, "%d", os.Getpid())
+	if err != nil {
+		err = fmt.Errorf("%s: could not write pid file, %v", pidFilePath, err)
+		logger.Error(err)
+		return err
+	}
+	err = fout.Close()
+	if err != nil {
+		err = fmt.Errorf("%s: could not close pid file, %v", pidFilePath, err)
+		logger.Error(err)
+		return err
+	}
+	return err
+}
 
-		// If we auto-loaded configs but a user supplied them as well, we have an error
-		if indexerConfigFound {
-			if configFile != "" {
-				logger.Errorf(
-					"indexer configuration was found in data directory (%s) as well as supplied via command line.  Only provide one.",
-					filepath.Join(indexerDataDir, autoLoadIndexerConfigName))
-				panic(exit{1})
-			}
+func runDaemon(daemonConfig *daemonConfig) error {
+	var err error
+	config.BindFlagSet(daemonConfig.flags)
+	err = configureLogger()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "failed to configure logger: %v", err)
+		return err
+	}
 
-			configFile = filepath.Join(indexerDataDir, autoLoadIndexerConfigName)
-			fmt.Printf("Auto-loading indexer configuration found: %s\n", configFile)
-		}
+	// Create the data directory if necessary/possible
+	if err = configureIndexerDataDir(daemonConfig.indexerDataDir); err != nil {
+		return err
+	}
 
-		if configFile != "" {
-			configs, err := os.Open(configFile)
-			if err != nil {
-				maybeFail(err, "error with config file (%s): %v", configFile, err)
-			}
-			defer configs.Close()
-			err = viper.ReadConfig(configs)
+	// Detect the various auto-loading configs from data directory
+	if err = loadIndexerConfig(daemonConfig.indexerDataDir, daemonConfig.configFile); err != nil {
```

Comment:
@AlgoStephenAkiki if we move this ahead of the `BindFlagSet` call, can we avoid calling that function twice?

Reply:
I don't think so since the problem is that we have a circular dependency on the logger inside the `loadIndexerConfig` function (the logger needs to be configured from configuration and the function uses the logger but also can re-load the configuration).